File: src/main/java/editor/database/symbol/PhyrexianSymbol.java
package editor.database.symbol;

import java.util.Arrays;
import java.util.Collections;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

import editor.database.attributes.ManaType;

/**
 * This class represents a Phyrexian mana symbol, which can be paid for either with mana of the
 * same color or two life.
 *
 * @author Alec Roelke
 */
public class PhyrexianSymbol extends ManaSymbol
{
    /**
     * Map of colors onto their corresponding Phyrexian symbols.
     */
    public static final Map<ManaType, PhyrexianSymbol> SYMBOLS = Collections.unmodifiableMap(
            Arrays.stream(ManaType.colors()).collect(Collectors.toMap(Function.identity(), PhyrexianSymbol::new)));

    /**
     * Get the Phyrexian symbol corresponding to the given String, which should be
     * a color character followed by either /p or /P.
     *
     * @param col String to parse
     * @return the corresponding Phyrexian symbol
     * @throws IllegalArgumentException if the String doesn't correspond to a Phyrexian symbol
     */
    public static PhyrexianSymbol parsePhyrexianSymbol(String col) throws IllegalArgumentException
    {
        PhyrexianSymbol symbol = tryParsePhyrexianSymbol(col);
        if (symbol == null)
            throw new IllegalArgumentException('"' + col + "\" is not a Phyrexian symbol");
        return symbol;
    }

    /**
     * Get the Phyrexian symbol corresponding to the given String, which should be
     * a color character followed by either /p or /P.
     *
     * @param col String to parse
     * @return the corresponding Phyrexian symbol, or null if there is none
     */
    public static PhyrexianSymbol tryParsePhyrexianSymbol(String col)
    {
        if (col.length() == 3 && col.charAt(1) == '/' && Character.toUpperCase(col.charAt(2)) == 'P')
            return SYMBOLS.get(ManaType.tryParseManaType(col.charAt(0)));
        return null;
    }

    /**
     * This PhyrexianSymbol's color.
     */
    public final ManaType color;

    /**
     * Create a new PhyrexianSymbol.
     *
     * @param color the new PhyrexianSymbol's color
     */
    private PhyrexianSymbol(ManaType color)
    {
        super("phyrexian_" + color.toString().toLowerCase() + "_mana.png", String.valueOf(color.shorthand()) + "/P", 1);
        this.color = color;
    }

    /**
     * {@inheritDoc}
     * This PhyrexianSymbols color weights are 0.5 for its color and 0 for all of the others.
     */
    @Override
    public Map<ManaType, Double> colorWeights()
    {
        return createWeights(new ColorWeight(color, 0.5));
    }

    @Override
    public int compareTo(ManaSymbol o)
    {
        if (o instanceof PhyrexianSymbol)
            return color.colorOrder(((PhyrexianSymbol)o).color);
        else
            return super.compareTo(o);
    }
}

File: src/main/scala/editor/database/symbol/PhyrexianSymbol.scala
package editor.database.symbol

import editor.database.attributes.ManaType

import scala.jdk.CollectionConverters._
import scala.jdk.OptionConverters._

/**
 * Companion to [[PhyrexianSymbol]] that contains all the possible symbols and methods for parsing them from strings.
 * @author Alec Roelke
 */
object PhyrexianSymbol {
  @deprecated val SYMBOLS = PhyrexianSymbolGenerator.values.asJava
  @deprecated def tryParsePhyrexianSymbol(s: String) = PhyrexianSymbolGenerator.parse(s).toJava
  @deprecated def parsePhyrexianSymbol(s: String) = PhyrexianSymbolGenerator.values(ManaType.parseManaType(s))
}

/**
 * A "Phyrexian" mana symbol that can be paid either with a color of mana or 2 life. Its color intensity is 0.5 for its color of mana, and 0 for
 * the rest, as there are two ways to pay for it (its color and 2 life), and its text representation is the shorthand of its [[ManaType]] and a P
 * separated by a slash ("/").
 * 
 * @constructor create a new Phyrexian mana symbol
 * @param color color of the new symbol
 * 
 * @author Alec Roelke
 */
class PhyrexianSymbol private[symbol](private val color: ManaType) extends ManaSymbol(s"phyrexian_${color.toString.toLowerCase}_mana.png", s"${color.shorthand.toUpper}/P", 1) {
  override def colorIntensity = ManaSymbol.createIntensity(Map(color -> 0.5))

  override def compareTo(o: ManaSymbol) = o match {
    case p: PhyrexianSymbol => color.compareTo(p.color)
    case _ => super.compareTo(o)
  }
}
